File: Data_structures/2d_Array.js
/**
 * @name 2D Array
  
 * @description @namespace {
        Like a 1D array, a 2D array is a collection 
        of data cells, all of the same type, which can 
        be given a single name. However, a 2D array is 
        organized as a matrix with a number of 
        rows and columns.
   }

 * @credits {
       https://computersciencewiki.org/index.php/Two-dimensional_arrays
   }
 */
let twoDArray = new Array(2); 
for(let i = 0; i<twoDArray.length; i++)
    twoDArray[i] = new Array(3);

let val, row, data, idx;
/**
 * @name Traverse
   
 * @algorithm {
     1.FOR i is 0 to row length;
       (Traverse each row)
       DO step#2;

     3.FOR j is 0 to column length;
       (Traverse each position in row a.k.a column)
       DO step#3;

     2.SET & ask user for data to fill 2*3 2d array;
       ADD user data to array;
    }
 */
for(let i = 0; i<twoDArray.length; i++) {

    for(let j = 0; j<twoDArray[i].length; j++) {
        val = parseFloat(prompt("Enter a number"));
        twoDArray[i][j] = val;
    }
}

console.log(twoDArray);

/** 
 * @name Insert
   
 * @algorithm {
     1.SET & ask user for data, row & position for insertion;

     2.IF row & position are in bounds;
       DO step#3;

     3.SHIFT all elements right of pos to right;
        (Overwriting value of pos in the process)
       INSERT data into pos after shifting;   
   }
*/
data = parseFloat(prompt("Enter a number"));
row = parseFloat(prompt("Enter row you want to insert " + data));
pos = parseFloat(prompt("What position in row " + row + " would you like to store "+ data))

if(pos <= 0 || row <= 0) {
    alert("Whoops, invalid row or position");
} else {
    let i = row-1;
    let len = twoDArray[i].length;
    for(let j = len-1; j>=pos-1; j--) {
        twoDArray[i][j+1] = twoDArray[i][j];
    }
    twoDArray[i][pos-1] = data;
}

console.log(twoDArray);

/** 
 * @name Delete
   
 * @algorithm {
     1.SET & ask user for row & position for deletion;

     2.IF row & position are in bounds;
       DO step#3;

     3.SHIFT all elements right of pos to left;
       REMOVE last element from row after shifting;   
   }
*/
row = parseFloat(prompt("Enter the row that you want to delete from"));
pos = parseFloat(prompt("Enter the position of the data you want to delete"));

if(row <= 0 || pos <= 0) {
    alert("Whoops, invalid row or position");
} else {
    let i = row-1;
    let len = twoDArray[i].length;
    let data = twoDArray[i][pos-1];
    for(let j = pos-1; j<len; j++) {
        twoDArray[i][j] = twoDArray[i][j+1];
    }
    alert("You've successfully removed " + data + " from array " + row + " at position " + pos);
    twoDArray[i][len-1] = null;
    idx = twoDArray[i].indexOf(null);
    twoDArray[i].splice(idx, 1);
}

console.log(twoDArray);

File: Data_structures/2d_Array.rb
# 2D Array

=begin
Like a 1D array, a 2D array is a collection 
of data cells, all of the same type, which can 
be given a single name. However, a 2D array is 
organized as a matrix with a number of 
rows and columns.

https://computersciencewiki.org/index.php/Two-dimensional_arrays
=end

# Create & Traverse
two_d_array = Array.new(2) {Array.new(3)}; 

i = 0;
until i === 2
    j = 0;
    until j === 3 
# get & ask user for data to fill 2*3 2d array.
        puts "Enter values for 2d array"
        val = gets.to_i;
# fill array with user data
        two_d_array[i][j] = val;
     j+=1;
    end
 i+=1;    
end

p two_d_array;

# Insert

# get & ask user for value, row & position/column.
puts "Please enter the row you want to insert your data at"
row = gets.to_i;
puts "Please enter a value";
val = gets.to_i;
puts "Please enter the position you want to store #{val}"
pos = gets.to_i;

if(row <= 0 || pos <= 0)
    puts "Whoops, invalid row or position"
else
    i = row-1;
    j = two_d_array[i].length-1;
    until(j === pos-2)
        two_d_array[i][j+1] = two_d_array[i][j];
     j-=1;
    end
    two_d_array[i][pos-1] = val;
end

p two_d_array;

# Delete

# get & ask user for row, postion/column to delete.
puts "Enter the row of the element you want to delete"
row = gets.to_i;
puts "Enter the position of the value you want to delete"
pos = gets.to_i;

i = row-1;
j = pos-1;
len = two_d_array[i].length;
until j === len
    two_d_array[i][j] = two_d_array[i][j+1];
 j+=1;   
end
two_d_array[i][len-1] = nil;
two_d_array[i].compact!();

p two_d_array